[1주차 기본 과제] 하찮은 가계부

ExceptAnyone opened the pull request:
<!-- PR의 제목은 "[ n주차 기본/심화/생각 과제 ] 과제명 " 으로 작성해주시면 되겠습니다 -->
<!-- PR은 리뷰어를 위한 글입니다, 보다 더 상세하게 적음으로써 소통을 활성화해봐요! -->

## ✨ 구현 기능 명세

- 기본 과제
1. **제목**
    
    > “OO이의 가계부”를 상단 제목으로 넣어줍니다.
    > 

1. **나의 자산**
    
    > 현재 나의 자산을 볼 수 있는 부분입니다.
    > 
    - 나의 자산 박스 위치와 내부 요소들 모두 중앙 정렬 되도록 해주세요.
    - 아래쪽에는 오늘의 총 수입과 총 지출이 나와있습니다. 총 수입과 총 지출의 색깔을 구분해주세요.  (ex. 수입 : 파란색, 지출 : 빨간색)

1. **날짜**
    - 좌 우로 누를 수 있는 버튼과, 그 가운데 오늘의 날짜가 있습니다.
    - 오늘의 날짜는 실제 날짜가 아닌, 임의로 지정하시면 됩니다.

1. **수입, 지출 선택**
    
    > 수입 선택시 수입만  지출 선택시 지출만 필터링되는 버튼입니다.
    > 
    - input type=”checkbox”로 구현합니다. (기본 스타일도 괜찮습니다.)
        
     
    - default는 수입, 지출 모두 checked 되어 있습니다.
    
2. **내역 리스트**
    



    > 수입 및 지출 내역 리스트입니다.
    > 
    - 리스트는 `flex`로 배치합니다.
    - 각 리스트에는 `카테고리` `내역 내용` `지출 및 수입 금액` `삭제 버튼 x` 가 있습니다.
    - 리스트 내부도 `flex`로 배치합니다. (선택)
    - 리스트가 일정 height 초과시 리스트 영역 내에서 스크롤이 생깁니다.
    - 지출과 수입 금액은 색깔 구분이 됩니다. (ex. 수입 : 파란색, 지출 : 빨간색)
    
3. **리스트 추가 버튼**
    
    > 최하단 가운데 추가 버튼이 있습니다.
    >

<br />

## 💎 PR Point
- ul태그 바깥과 안에 div태그를 따로 설정해두었는데, 굳이 이렇게 할 필요가 있나 생각이 듭니다...
```html
<div class="pay_detail_contents">
        <ul>
          <div class="pay_detail_item">
            <li>
              <p class="pay_detail_category">식비</p>
              <p class="pay_detail_content">맥도날드</p>
              <p class="pay_detail_amount_spend">-10,000</p>
              <button class="pay_detail_btn">x</button>
            </li>
          </div>
```
<br />

- 부분 스크롤을 구현하기 위해 overflow -x : hidden을 사용하였습니다.
```css
.pay_detail_contents {
  width: 100%;
  height: 500px;
  overflow-x: hidden;
}
```
- 이렇게 하였더니, 풀 스크린에서는 정상적으로 출력되는 x 버튼이, 
화면의 크기를 줄였더니 사라지더라구여. 추후 리팩토링 필요할 것 같습니다.

- 하단 + 버튼을 sticky로 고정하였습니다.
```css
.footer {
  position: sticky;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
}
```

## 🥺 소요 시간, 어려웠던 점

- `3h`
- grid를 다루었던 이전과제보다는 괜찮았다고 느껴집니다. 
- 저는 css에 전혀 재능이 없다는 것을 절실히 깨닫게되었습니다. 
- 제 결과물은 너무 안이쁘네요

<br />

## 🌈 구현 결과물

 https://github.com/DO-SOPT-WEB/JeonganJang/assets/126966681/ef91efa4-8c22-4940-82ff-8baa27dd6f31

ExceptAnyone committed:
init: 초기세팅

ExceptAnyone committed:
feat: 헤더 생성

ExceptAnyone committed:
chore: img/video 파일 추가

ExceptAnyone committed:
chore: 이미지 파일 추가 업로드

ExceptAnyone committed:
feat: 헤더 설정 및 가운데 이미지 가로스크롤, nav바 설정

헤더를 검정색 바탕에 흰색 글씨로 설정하고, 화면 중간 섹션에 가로스크롤을 구현하였으며,
nav_list를 두고 hover시 색상변경과 밑줄이 그어지게 하였습니다

ExceptAnyone committed:
chore: 필요없는 이미지 삭제

ExceptAnyone committed:
style: nav_title 중간 이동

ExceptAnyone committed:
nav_list에 a태그 추가

ExceptAnyone committed:
style: 링크에 밑줄과 색상 제거

ExceptAnyone committed:
feat: 목차 클릭시 해당 섹션으로 이동

ExceptAnyone committed:
style: 첫 컨텐츠 섹션 헤더 부분 상단 고정 및 해당 이미지 사이즈 조절

ExceptAnyone committed:
chore: img 파일 추가

음식 사진이 부족하여 관련 이미지 파일을 추가하였습니다.

ExceptAnyone committed:
feat: 두 번째, 세 번째 섹션 기본틀 구현
(추후 flex,grid 이용하여 배치 예정)

추후에 flex,grid 이용하여 배치하겠습니다. 우선적으로 사진만 배치 해놓았습니다.

ExceptAnyone committed:
feat: 두 번째 섹션 배치 구현

ExceptAnyone committed:
feat: footer 설정 및 상단 이동 버튼 생성

세 번째 섹션인 grid 관련해서 일부 주석처리 후 다시 작업 예정입니다.

ExceptAnyone committed:
style: 세 번째 섹션 grid형식 배치

ExceptAnyone committed:
Feat: 헤더 세팅 및 첫 번째 섹션 구현

헤더를 세팅한 후, 나의 자산 부분인 첫 번째 섹션까지 구현하였습니다

ExceptAnyone committed:
feat: 날짜 및 수입,지출 선택 버튼

ExceptAnyone committed:
feat: 내역 리스트 세부 내역 작성 및 스타일링, footer & 하단 버튼 구현

moondda commented on week1/style.css:
```diff
@@ -0,0 +1,162 @@
+  text-decoration: underline;
+}
+
+.first_content_header {
+  position: sticky;
+  top: 0; /* 화면 맨 위에 고정 */
+  background: skyblue;
+  padding: 10px;
+  justify-content: center;
+  z-index: 100; /* 다른 요소 위에 표시 */
+}
```
first_content_header랑 second_content_header랑 css내용이 완전 똑같넹 이런 경우에는 그냥 class 하나로 통일하는게 더 좋을듯??

moondda commented on week1-2/style.css:
```diff
@@ -0,0 +1,141 @@
+  gap: 50px;
+}
+
+.amount_detail_plus_container,
+.amount_detail_minus_container {
+  text-align: center;
+  display: flex;
+}
+
+.amount_detail_plus_container {
+  color: blue;
+}
+
+.amount_detail_minus_container {
+  color: red;
+}
```
나는 이렇게 css가 하나씩 다른 경우에는 그냥 인라인스타일로 정의해주는 거 같아!! 그게 특정 영역에 스타일 적용할 때 가장 빠르니까?? 하지만 로딩시간이 더 오래걸리고 유지보수가 어렵다고 하네 인라인스타일은...모든지 트레이드오프가 있으니 적절히 사용하면 될듯ㅎ

ExceptAnyone replied:
아 맞슴다!! 다현이 너꺼 코드보면서 딱 처음에 든 생각이 '아 나도 저렇게 간결하게 할껄' 생각 들더라고 !! 덕분에 많이 배웠음 ㅎ.ㅎ

Arooming replied:
마자! 이 부분은 다현이가 잘 짚어준 것 같아! 👍👍🤍
인라인 스타일은 특정 영역에서 직관적이라 간단한 테스트/ 개발 시에 많이 쓰인다고 해! 
지금은 꽤 간단한 코드를 작성하고 있으니, 가끔은 필요 시 선택적으로 사용해도 될 것 같긴하당

그치만 인라인 스타일은 우선 순위가 가장 높아서, 호옥시라도 인라인 스타일 적용한 걸 까먹고 나중에 내/외부 스타일 시트에서 해당 속성 값을 변경하려고 시도하는 경우에는 (....끔찍🥶...) 계속 헤매게 될 수도 있으니까 인라인 스타일 적용은 최대한 지양해주는게 좋긴 하겠다 !! 🤑

ps. 아주 개인적인 의견을 하나 말해보자면, 나는 습관을 잘 들이는 것도 중요하다고 생각해서 지양해야하는 부분들은 최대한 배제하고 코드를 구현하는 편이긴해 !! ㅎㅎ

seobbang replied:
맞아용~! 아름언니 말처럼 인라인 스타일은 간편하다는 특징이 있지만! 결국 유지보수 측면에 좋지 않기 때문에 지양하는 편이 좋습니당 ㅎㅎ

Arooming commented on week1-2/index.html:
```diff
@@ -0,0 +1,103 @@
+    </header>
+    <main>
+      <section class="all_my_amount">
+        <div class="my_amount">
+          <div class="my_amount_title">나의 자산</div>
+          <div class="my_amount_sub">300,000</div>
+        </div>
+        <div class="amount_detail_container">
+          <div class="amount_detail_plus_container">
+            <p class="amount_detail_plus">+</p>
+            <p class="amount_detail_plus">600,000</p>
+          </div>
+          <div class="amount_detail_minus_container">
+            <p class="amount_detail_minus">-</p>
+            <p class="amount_detail_minus">50,300</p>
+          </div>
+        </div>
```
여기 `div`가 많이 쓰인 것 같은데..!! 최대한 시멘틱하게 바꿔보면 어떨까 !!
개인적인 생각으로는 `.my_account`는 `나의 자산`이라는 주제를 묶는 독립적인 컨텐츠라는 생각이 들어서, `article` 태그로 바꿔봐도 좋을 것 같다 !

그리고 div 태그는 아무 의미없는 컨테이너 역할을 한다고 알고 있는데, 요런 의미에서 `나의 자산`과 `300,000`을 div 대신 문맥을 구분하는 역할인 `p` 태그로 바꿔주는건 어떤 것 같아!?!?!! 
오빠의 생각이 궁금합니다!!

ExceptAnyone replied:
안그래도 div가 많아서 토할 뻔했는데 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ p로 바꾸는게 훨씬 좋은 것 같다

Arooming commented on week1-2/index.html:
```diff
@@ -0,0 +1,103 @@
+        <div class="pay_detail_list">
+          <p>내역 리스트</p>
+          <form>
+            <input type="checkbox" name="income" value="income" checked />수입
+            <input type="checkbox" name="spend" value="spend" checked />지출
```
`<input>` 과 `<label>`을 연결해서 구현하는 방법도 있어!
이런 식으로 하면 체크박스를 누르지 않고 관련 `<label>`을 클릭하더라도 `<input>` 자체에 초점을 맞추거나 활성화 시킬 수 있으니까 `<label>` 활용한 코드로 바꿔봐도 좋을듯!

[`<label>` 관련 참고 링크](https://developer.mozilla.org/ko/docs/Web/HTML/Element/label)

Arooming commented on week1-2/index.html:
```diff
@@ -0,0 +1,103 @@
+          </div>
+        </div>
+      </section>
+      <section class="pay_detail">
+        <div class="date">
```
아래 css 코드도 보고 왔는데 웬만한 속성은 section에 거의 다 줘서 div는 사실상 거의 맡은 역할이 없는 것 같은데, 혹시 요런 식으로 나눠준 이유가 있을까?!

ExceptAnyone replied:
이유가 있는건 아니고... 습관처럼 div를 쓴 느낌?? 이런 습관 지양해야될 것 같아 ㅋㅋㅋ

Arooming commented on week1-2/index.html:
```diff
@@ -0,0 +1,103 @@
+          </form>
+        </div>
+      </article>
+      <div class="pay_detail_contents">
```
요기 저번에 오빠가 내 코드에 리뷰 남겨준 것처럼, ul 태그 활용해서 시멘틱하게 잘 만들어줬으니까 div 태그는 빼고 div에 적용한 css 내용을 ul 태그에 넣어주는 방향으로 리팩토링 해보장!!!

ExceptAnyone replied:
완료!

Arooming commented on week1-2/index.html:
```diff
@@ -0,0 +1,103 @@
+      </article>
+      <div class="pay_detail_contents">
+        <ul>
+          <div class="pay_detail_item">
+            <li>
```
요기랑 아래 li 태그들 모두 !! div 태그는 빼고 해당 속성을 li 태그에 적용하는 방식으로 바꾸기 !! 💪💪

ExceptAnyone replied:
굳굳!! 완료

Arooming commented on week1-2/style.css:
```diff
@@ -0,0 +1,141 @@
+  width: 100%;
+}
+
+.my_amount_title {
+  font-size: 50px;
+  margin-bottom: 5px;
+  border-bottom: 8px solid #4390e1;
+}
```
요기도 앞에 과제랑 마찬가지로 상대 단위인 `rem`으로 수정해보는거 어때 !!

ExceptAnyone replied:
모든 px단위를 rem으로 수정해보자는 말이징??

Arooming replied:
우웅 아무래도 px은 고정적인 값이라 화면 크기나 폰트 크기가 변해도 그대로 유지되기 때문에 반응형 대응이 힘들어서!
최대한 지양하는게 좋다고 생각해!

Arooming commented on week1-2/style.css:
```diff
@@ -0,0 +1,141 @@
+li {
+  display: flex;
+  justify-content: space-between;
+  margin: 1%;
```
[margin과 padding 값을 백분율](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Sizing_items_in_CSS)로 설정하면 원하는 결과가 나오지 않을 수도 있어서 지양하는게 좋다고 해!

margin과 padding은 모두 [컨테이닝 블록](https://developer.mozilla.org/ko/docs/Web/CSS/Containing_block)의 `너비!!`를 기준으로 하기 때문에, 조상 요소의 너비를 잘 파악해서 % 값을 줬음에도 **컨테이닝 블록 !== 조상 요소**(조상요소가 꼭 컨테이닝 블록이 되는 것은 아니기에 ....) 라면 원하는 결과가 나오지 않을 수 도 있대! 🫢

걸어둔 링크들 꼭꼭 한 번 읽어보기 !

ExceptAnyone replied:
아 이부분 시간에 쫓겨서 과제하느라 별 생각없이 작성했던 것 같아 ㅠ 링크 고마워 방금 다 읽었어 ㅎㅎ

Arooming commented on week1-2/style.css:
```diff
@@ -0,0 +1,141 @@
+.pay_detail_contents {
+  width: 100%;
+  height: 500px;
+  overflow-x: hidden;
```
pr 포인트에 적어줬던 화면의 너비가 줄어들 때 x 버튼이 사라지는 이유는 ul 태그에 `overflow-x: hidden;`를 줘서 인 것 같아!
이렇게 되면 일정 길이가 넘어갔을 때, 내용이 넘치는 부분이 화면에서 잘리게 되거든 ..!!
아마 화면 크기는 줄어들었는데 요소의 크기는 그에 비례하게 줄어들지 않아서 발생한 문제인 것 같다는 생각이 드네용 ..! 

목록 부분 스크롤을 주고 싶다면, `overflow-y: scroll;`을 활용해보기 !

ExceptAnyone replied:
어라 내가 처음에 `overflow-y: scroll;` 했을 때는 안되었는디 지금 하니까 바로 되네? ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

Arooming reviewed (commented):
두번째 과제도 너무너무 수고했다!! 와아아아아👏
전반적으로 css 단위 통일해주고 시맨틱 태그 부분만 더 신경써주면 완전 괜찮을 것 같은데 !?!??!?!
2주차 과제도 아좌좟! 과제 두 개 다 넘 잘했다잉 ~ 👍

ExceptAnyone committed:
refactor: 코드리뷰 반영

div태그가 많은부분 시맨틱하게 변경
px단위에서 rem단위로 변경(추후 다시)
좁아진 화면에서 x버튼이 사라지는 원인 수정
(overflow-y:scroll)

금잔디 1조분들 자세하게 리뷰 남겨주셔서 감사합니다!! 덕분에 많이 배워가요!!!

seobbang commented on week1-2/style.css:
```diff
@@ -0,0 +1,137 @@
+  font-weight: bold;
+}
+
+.pay_detail_amount_spend {
+  color: red;
+}
+
+.pay_detail_amount_income {
+  color: blue;
+}
```
위에 있는 `amount_detail_plus_container`와 `amount_detail_minus_container` 그리고 이 친구들은 전부 color 스타일링 만을 위해서 존재하는 것 같은데, 이런 경우에는 두가지 방법이 있을 것 같아!
1️⃣ 한번에 쓰기!
```suggestion
.pay_detail_amount_spend, amount_detail_minus_container {
  color: red;
}

.pay_detail_amount_income, amount_detail_plus_container {
  color: blue;
}
```

2️⃣ 아래처럼 줘서 각 색마다 같은 클래스를 주기
```suggestion
.spend {
  color: red;
}

.income {
  color: blue;
}
```

seobbang reviewed (commented):
굿굿 리팩토링 할 수 있는 사항들 꼭 반영해보쟈!! 🔥
